src/bech32.cpp

```diff
@@ -31,182 +34,73 @@ const int8_t CHARSET_REV[128] = {
-    191, 217, 847, 941, 424, 235, 38, 74, 616, 786, 147, 4, 273, 214, 142,
-    575, 992, 463, 983, 243, 360, 970, 350, 267, 615, 766, 494, 31, 1009,
-    452, 710, 552, 128, 612, 600, 275, 322, 193
-};
-static_assert(std::size(GF1024_LOG) == 1024, "GF1024_EXP length should be 1024");
+/** We work with the finite field GF(1024) defined as a degree 2 extension of the base field GF(32)
+ * The defining polynomial of the extension is x^2 + 9x + 23.
+ * Let (e) be a root of this defining polynomial. Then (e) is a primitive element of GF(1024),
+ * that is, a generator of the field. Every non-zero element of the field can then be represented
+ * as (e)^k for some power k.
+ * The array GF1024_EXP contains all these powers of (e) - GF1024_EXP[k] = (e)^k in GF(1024).
+ * Conversely, GF1024_LOG contains the discrete logarithms of these powers, so
+ * GF1024_LOG[GF1024_EXP[k]] == k.
+ * The following function generates the two tables GF1024_EXP and GF1024_LOG as constexprs. */
+constexpr std::pair<std::array<int16_t, 1023>, std::array<int16_t, 1024>> generate_gf_tables()
+{
+    // Build table for GF(32).
+    // We use these tables to perform arithmetic in GF(32) below, when constructing the
+    // tables for GF(1024).
+    std::array<int8_t, 31> GF32_EXP{};
+    std::array<int8_t, 32> GF32_LOG{};
+
+    // fmod encodes the defining polynomial of GF(32) over GF(2), x^5 + x^3 + 1.
+    // Because coefficients in GF(2) are binary digits, the coefficients are packed as 101001.
+    const int fmod = 41;
+
+    GF32_EXP[0] = 1;
+    GF32_LOG[0] = -1;
+    GF32_LOG[1] = 0;
+    int v = 1;
+    for (int i = 1; i < 31; ++i) {
+        v = v << 1;
+        if (v & 32) v ^= fmod;
+        GF32_EXP[i] = v;
+        GF32_LOG[v] = i;
+    }
+
+    // Build table for GF(1024)
+    std::array<int16_t, 1023> GF1024_EXP{};
+    std::array<int16_t, 1024> GF1024_LOG{};
+
+    GF1024_EXP[0] = 1;
+    GF1024_LOG[0] = -1;
+    GF1024_LOG[1] = 0;
+
+    // Each element v of GF(1024) is encoded as a 10 bit integer in the following way:
+    // v = v1 || v0 where v0, v1 are 5-bit integers (elements of GF(32)).
+    // The element (e) is encoded as 1 || 0, to represent 1*(e) + 0. Every other element
+    // a*(e) + b is represented as a || b (a,b are both GF(32) elements). Given (v), we
+    // compute (e)*(v) by multiplying in the following way:
+    // v0' = 23*v1
+    // v1' = 9*v1 + v0
+    // e*v = v1' || v0'
+    v = 1;
+    for (int i = 1; i < 1023; ++i) {
+        int v0 = v & 31;
+        int v1 = v >> 5;
+
+        int v0n = v1 ? GF32_EXP.at((GF32_LOG.at(v1) + GF32_LOG.at(23)) % 31) : 0;
+        int v1n = (v1 ? GF32_EXP.at((GF32_LOG.at(v1) + GF32_LOG.at(9)) % 31) : 0) ^ v0;
+
+        v = v1n << 5 | v0n;
+        GF1024_EXP[i] = v;
+        GF1024_LOG[v] = i;
+    }
+
+    return std::make_pair(GF1024_EXP, GF1024_LOG);
+}
+
+constexpr auto tables = generate_gf_tables();
+constexpr std::array<int16_t, 1023> GF1024_EXP = tables.first;
```

Comment:
I *assume* this doesn't actually materialize the tables twice, as one would be an unused symbol in an anonymous namespace. Still, if you'd want to avoid the copying at compile-time (or rule out the risk that it gets emitted twice in the binary), these `GF1024_EXP` and `GF1024_LOG` variables could be references (e.g. `constexpr const std::array<int16_t, 1023>&`).